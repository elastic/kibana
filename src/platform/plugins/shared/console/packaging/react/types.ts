/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

// Standalone type definitions, here we avoid imports to avoid compiling dependencies

export interface HttpSetup {
  get: <T = any>(url: string, options?: { query?: Record<string, any> }) => Promise<T>;
  post: <T = any>(url: string, body?: any, options?: { query?: Record<string, any> }) => Promise<T>;
  delete: <T = any>(url: string, options?: { query?: Record<string, any> }) => Promise<T>;
  put: <T = any>(url: string, body?: any, options?: { query?: Record<string, any> }) => Promise<T>;
  patch: <T = any>(
    url: string,
    body?: any,
    options?: { query?: Record<string, any> }
  ) => Promise<T>;
  head: <T = any>(url: string, options?: { query?: Record<string, any> }) => Promise<T>;
  fetch: <T = any>(
    url: string,
    options?: { method?: string; body?: any; query?: Record<string, any> }
  ) => Promise<T>;
}

// Toast notification types
export interface ToastInput {
  title?: string;
  text?: string;
  iconType?: string;
  color?: 'primary' | 'success' | 'warning' | 'danger';
}

export interface NotificationCallbacks {
  addSuccess?: (toastOrTitle: ToastInput | string) => void;
  addWarning?: (toastOrTitle: ToastInput | string) => void;
  addDanger?: (toastOrTitle: ToastInput | string) => void;
  addError?: (error: Error, options?: { title?: string }) => void;
  add?: (toast: ToastInput) => { id: string };
  remove?: (toastId: string) => void;
}

export interface OneConsoleProps {
  lang?: 'en' | 'fr-FR' | 'ja-JP' | 'zh-CN';
  http?: Partial<HttpSetup>;
  notifications: NotificationCallbacks;
}
