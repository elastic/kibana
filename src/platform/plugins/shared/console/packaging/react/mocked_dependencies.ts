/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

import { createMockedCoreContext } from '@kbn/standalone-packaging-utils';
import { createMockedInjectedMetadata } from '@kbn/standalone-packaging-utils';
import { createMockedTrackUiMetric } from '@kbn/standalone-packaging-utils';

export const coreContext = createMockedCoreContext({
  version: '1.0.0',
  branch: 'main',
  buildNum: 1,
  buildSha: 'prod-build',
  buildShaShort: 'prod',
  mode: 'production',
  enableConsoleLogging: true,
});

export const injectedMetadata = createMockedInjectedMetadata({
  kibanaBranch: 'main',
  kibanaVersion: '9.3.0',
  kibanaBuildNumber: 12345,
});

export const trackUiMetric = createMockedTrackUiMetric();
