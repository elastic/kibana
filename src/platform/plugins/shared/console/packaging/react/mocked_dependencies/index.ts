/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

export { injectedMetadata } from './injected_metadata';
export { coreContext } from './core_context';
export { trackUiMetric } from './track_ui_metric';
