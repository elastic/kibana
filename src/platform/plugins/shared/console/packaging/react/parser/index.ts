/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

import { createStandaloneParsedRequestsProvider } from './standalone_console_parser';

// Create a provider factory for packaging environment
export const createPackagingParsedRequestsProvider = () => {
  return (model: any) => {
    if (!model) {
      // eslint-disable-next-line no-console
      console.warn('Monaco editor model is null, creating fallback provider');
      return {
        getRequests: () => Promise.resolve([]),
        getErrors: () => Promise.resolve([]),
      };
    }
    return createStandaloneParsedRequestsProvider(model);
  };
};
