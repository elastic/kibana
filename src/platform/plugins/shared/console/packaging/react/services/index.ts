/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

// Export factory functions and types
export { createInjectedMetadata, type InjectedMetadataOptions } from './injected_metadata';
export { createCoreContext, type CoreContextOptions } from './core_context';
export { createTrackUiMetric, type TrackUiMetricOptions } from './track_ui_metric';

// Import factory functions for service initialization
import { createCoreContext } from './core_context';
import { createInjectedMetadata } from './injected_metadata';
import { createTrackUiMetric } from './track_ui_metric';

// Initialize and export service instances
export const coreContext = createCoreContext({
  version: '1.0.0',
  branch: 'main',
  buildNum: 1,
  buildSha: 'prod-build',
  buildShaShort: 'prod',
  mode: 'production',
  enableConsoleLogging: true,
});

export const injectedMetadata = createInjectedMetadata({
  kibanaBranch: 'main',
  kibanaVersion: '9.3.0',
  kibanaBuildNumber: 12345,
});

export const trackUiMetric = createTrackUiMetric();
