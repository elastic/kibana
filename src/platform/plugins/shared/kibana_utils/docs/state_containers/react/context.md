# React context

`createStateContainerReactHelpers` returns `<Provider>` and `<Consumer>` components
as well as `context` React context object.

```ts
export const {
  Provider,
  Consumer,
  context,
} = createStateContainerReactHelpers<typeof container>();
```

`<Provider>` and `<Consumer>` are just regular React context components.

```tsx
<Provider value={container}>
  <div>
    <Consumer>{container =>
      <pre>{JSON.stringify(container.get())}</pre>
    }</Consumer>
  </div>
</Provider>
```
