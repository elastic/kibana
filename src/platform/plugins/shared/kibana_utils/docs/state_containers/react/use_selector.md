# `useSelector()` hook

With `useSelector` React hook you specify a selector function, which will pick specific
data from the state. *Your component will update only when that specific part of the state changes.*

```tsx
const selector = state => state.isDarkMode;
const Demo = () => {
  const isDarkMode = useSelector(selector);
  return <div>{isDarkMode ? '🌑' : '☀️'}</div>;
};
```

As an optional second argument for `useSelector` you can provide a `comparator` function, which
compares currently selected value with the previous and your component will re-render only if
`comparator` returns `true`. By default it uses [`fast-deep-equal`](https://github.com/epoberezkin/fast-deep-equal).

```
useSelector(selector, comparator?)
```
