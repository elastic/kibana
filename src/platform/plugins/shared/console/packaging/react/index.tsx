/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

// This is a packaged standalone version of console that needs to import browser code
/* eslint-disable @kbn/imports/no_boundary_crossing */

import React, { useEffect, useMemo, useState, useRef } from 'react';
import { noop } from 'lodash';

// Disable Monaco workers entirely, in case monaco ever decides to spawn any for some reason.
// Console doesnt use any of the monaco features that require workers, since we provide
// our own language tokenizer and autocompletion provider.
(window as any).MonacoEnvironment = {
  getWorker() {
    // Return a minimal mock worker that doesn't actually do anything
    return {
      postMessage: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      terminate: () => {},
      onerror: null,
      onmessage: null,
      onmessageerror: null,
    };
  },
};

import { __IntlProvider as IntlProvider } from '@kbn/i18n-react';

import { DocLinksService } from '@kbn/core-doc-links-browser-internal';
import type { CoreContext } from '@kbn/core-base-browser-internal';

import { HttpService } from '@kbn/core-http-browser-internal';
import { ExecutionContextService } from '@kbn/core-execution-context-browser-internal';
import { FatalErrorsService } from '@kbn/core-fatal-errors-browser-internal';
import { AnalyticsService } from '@kbn/core-analytics-browser-internal';
import { ThemeService } from '@kbn/core-theme-browser-internal';
import { I18nService } from '@kbn/core-i18n-browser-internal';
import { i18n } from '@kbn/i18n';
import { type HttpSetup } from '@kbn/core/public';
import type { NotificationsSetup } from '@kbn/core/public';
import { createStorage, createHistory, createSettings, setStorage } from '../../public/services';
import { loadActiveApi } from '../../public/lib/kb';

import { createPackagingParsedRequestsProvider } from './parser';
import { injectedMetadata, coreContext, trackUiMetric } from './services';

import * as localStorageObjectClient from '../../public/lib/local_storage_object_client';
import { Main } from '../../public/application/containers';

import {
  ServicesContextProvider,
  EditorContextProvider,
  RequestContextProvider,
} from '../../public/application/contexts';
import { createApi, createEsHostService } from '../../public/application/lib';
import { AutocompleteInfo, setAutocompleteInfo } from '../../public/services';
import type { OneConsoleProps } from './types';

// Translation files are generated during the build process
/* eslint-disable @kbn/imports/no_unresolvable_imports */
const translations = {
  en: {
    formats: {},
    messages: {},
  },
  'fr-FR': require('./translations/fr-FR.json'),
  'ja-JP': require('./translations/ja-JP.json'),
  'zh-CN': require('./translations/zh-CN.json'),
  'de-DE': require('./translations/de-DE.json'),
};
/* eslint-enable @kbn/imports/no_unresolvable_imports */

export const OneConsole = ({
  lang = 'en',
  http: customHttp,
  notifications: customNotifications,
}: OneConsoleProps) => {
  const [apiLoaded, setApiLoaded] = useState(false);

  // Get the translations for the selected language, fallback to English
  const selectedTranslations = translations[lang] || translations.en;

  // Configure the global @kbn/i18n system with the same translations
  i18n.init({
    locale: lang,
    formats: selectedTranslations.formats,
    messages: selectedTranslations.messages,
  });

  // Create all services once using useRef as they should never be recreated
  const servicesRef = useRef<any>(null);

  if (!servicesRef.current) {
    const docLinksService = new DocLinksService(coreContext as CoreContext);
    docLinksService.setup();
    const docLinks = docLinksService.start({ injectedMetadata });

    const i18nService = new I18nService();

    const themeService = new ThemeService();
    const theme = themeService.setup({ injectedMetadata });

    const analyticsService = new AnalyticsService(coreContext as CoreContext);
    const analytics = analyticsService.setup({ injectedMetadata });

    const rootDomElement = document.getElementById('root')!;
    const fatalErrorsService = new FatalErrorsService(rootDomElement, () => {
      // eslint-disable-next-line no-console
      console.log('FATAL ERROR OCURRED');
    });
    const fatalErrors = fatalErrorsService.setup({
      injectedMetadata,
      analytics,
      theme,
      i18n: i18nService.getContext(),
    });

    const executionContextService = new ExecutionContextService();
    const executionContext = executionContextService.setup({ analytics });

    const httpService = new HttpService();
    const originalHttp = httpService.setup({
      injectedMetadata,
      fatalErrors,
      executionContext,
    });

    const http = {
      ...originalHttp,
      ...customHttp,
    } as HttpSetup;

    const storage = createStorage({
      engine: window.localStorage,
      prefix: 'sense:',
    });
    setStorage(storage);

    const storageHistory = createHistory({ storage });
    const settings = createSettings({ storage });
    const objectStorageClient = localStorageObjectClient.create(storage);
    const api = createApi({ http });
    const esHostService = createEsHostService({ api });

    // Initialize autocompleteInfo like in the plugin
    const autocompleteInfo = new AutocompleteInfo();
    autocompleteInfo.setup(http);
    autocompleteInfo.mapping.setup(http, settings);
    // Initialize autocomplete
    setAutocompleteInfo(autocompleteInfo);

    servicesRef.current = {
      http,
      docLinks,
      theme,
      i18nService,
      storage,
      storageHistory,
      settings,
      objectStorageClient,
      esHostService,
      autocompleteInfo,
    };
  }

  const {
    http,
    docLinks,
    theme,
    i18nService,
    storage,
    storageHistory,
    settings,
    objectStorageClient,
    esHostService,
    autocompleteInfo,
  } = servicesRef.current;

  // Use the custom notifications provided by the consumer
  const notifications = useMemo(
    () => ({
      toasts: {
        addSuccess: customNotifications.addSuccess || noop,
        addWarning: customNotifications.addWarning || noop,
        addDanger: customNotifications.addDanger || noop,
        addError: customNotifications.addError || noop,
        add: customNotifications.add || noop,
        remove: customNotifications.remove || noop,
      },
    }),
    [customNotifications]
  );

  useEffect(() => {
    const loadApi = async () => {
      try {
        await loadActiveApi(http);
        setApiLoaded(true);
      } catch (error) {
        setApiLoaded(true);
      }
    };

    loadApi();
  }, [http]);

  // Don't render until API is loaded
  if (!apiLoaded) {
    return null;
  }

  return (
    <IntlProvider locale={lang} messages={selectedTranslations.messages}>
      <ServicesContextProvider
        value={{
          analytics: {
            reportEvent: () => {},
          },
          i18n: i18nService.getContext(),
          theme: {
            theme$: theme.theme$,
          },
          userProfile: {
            getUserProfile$: () => ({ pipe: () => ({ subscribe: () => {} }) }),
            userProfile: null,
          } as any,
          docLinkVersion: docLinks.DOC_LINK_VERSION,
          docLinks: docLinks.links,
          services: {
            esHostService,
            storage,
            history: storageHistory,
            settings,
            notifications: notifications as NotificationsSetup,
            trackUiMetric,
            objectStorageClient,
            http,
            autocompleteInfo,
            // Mock required services that aren't available in packaging
            dataViews: {} as any,
            data: {} as any,
            licensing: {} as any,
            application: {} as any,
          },
          config: {
            isDevMode: false,
            isPackagedEnvironment: true,
          },
        }}
      >
        <RequestContextProvider>
          <EditorContextProvider
            settings={settings.toJSON()}
            customParsedRequestsProvider={createPackagingParsedRequestsProvider()}
          >
            <div className="kbnConsole">
              <Main />
            </div>
          </EditorContextProvider>
        </RequestContextProvider>
      </ServicesContextProvider>
    </IntlProvider>
  );
};
