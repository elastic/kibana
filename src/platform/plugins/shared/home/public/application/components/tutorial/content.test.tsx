/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

import React from 'react';
import { render } from '@testing-library/react';
import { Content } from './content';

describe('Content component', () => {
  test('should render Content component with markdown', async () => {
    const { container } = render(
      <Content
        text={'I am *some* [content](https://en.wikipedia.org/wiki/Content) with `markdown`'}
      />
    );
    expect(container).toMatchInlineSnapshot(`
    <div>
      <div
        class="euiText euiMarkdownFormat emotion-euiText-m-euiTextColor-default-euiMarkdownFormat-m-default"
      >
        <p>
          I am 
          <em>
            some
          </em>
           
          <a
            class="euiLink emotion-euiLink-primary"
            href="https://en.wikipedia.org/wiki/Content"
            rel="noreferrer"
          >
            content
          </a>
           with 
          <code>
            markdown
          </code>
        </p>
      </div>
    </div>
    `);
  });
});
