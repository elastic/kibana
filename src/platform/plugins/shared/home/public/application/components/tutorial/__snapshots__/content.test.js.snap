// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render content with markdown 1`] = `
<EuiMarkdownFormat>
  I am *some* [content](https://en.wikipedia.org/wiki/Content) with \`markdown\`
</EuiMarkdownFormat>
`;
